File: app/src/main/java/com/huynd/skyobserver/utils/PriceComparator.java
package com.huynd.skyobserver.utils;

import com.huynd.skyobserver.models.PricePerDay;

import java.util.Comparator;

import static com.huynd.skyobserver.utils.PriceComparator.SortOrder.DEPART_EARLIEST;

/**
 * Created by HuyND on 10/11/2017.
 */

public class PriceComparator implements Comparator<PricePerDay> {

    private static class SingletonHolder {
        static PriceComparator sInstance = new PriceComparator();
    }

    public enum SortOrder {
        PRICE_ONLY_LOWEST,
        PRICE_ONLY_HIGHEST,
        TOTAL_PRICE_LOWEST,
        TOTAL_PRICE_HIGHEST,
        DEPART_EARLIEST,
        DEPART_LATEST,
        AIRLINES,
        UNKNOWN
    }

    private SortOrder mOrder = DEPART_EARLIEST;

    private PriceComparator() {}

    public static PriceComparator getInstance() {
        return SingletonHolder.sInstance;
    }

    @Override
    public int compare(PricePerDay price1, PricePerDay price2) {
        if (price1 == null && price2 == null) {
            return 0;
        }

        if (price1 == null) {
            return -1;
        }

        if (price2 == null) {
            return 1;
        }

        switch (mOrder) {
            case PRICE_ONLY_LOWEST:
                return price1.getPrice() - price2.getPrice();
            case PRICE_ONLY_HIGHEST:
                return price2.getPrice() - price1.getPrice();
            case TOTAL_PRICE_LOWEST:
                return price1.getPriceTotal() - price2.getPriceTotal();
            case TOTAL_PRICE_HIGHEST:
                return price2.getPriceTotal() - price1.getPriceTotal();
            case DEPART_EARLIEST:
                return price1.getDepartureTime().compareTo(price2.getDepartureTime());
            case DEPART_LATEST:
                return price2.getDepartureTime().compareTo(price1.getDepartureTime());
            case AIRLINES:
                return price1.getCarrier().compareTo(price2.getCarrier());
        }
        return 0;
    }

    public void setSortOrder(SortOrder order) {
        mOrder = order;
    }
}

File: app/src/main/java/com/huynd/skyobserver/utils/PriceComparator.kt
package com.huynd.skyobserver.utils

import com.huynd.skyobserver.entities.PricePerDay
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.AIRLINES
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.DEPART_EARLIEST
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.DEPART_LATEST
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.PRICE_ONLY_HIGHEST
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.PRICE_ONLY_LOWEST
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.TOTAL_PRICE_HIGHEST
import com.huynd.skyobserver.utils.PriceComparator.SortOrder.TOTAL_PRICE_LOWEST
import java.util.*

/**
 * Created by HuyND on 10/11/2017.
 */

class PriceComparator : Comparator<PricePerDay> {
    enum class SortOrder {
        PRICE_ONLY_LOWEST,
        PRICE_ONLY_HIGHEST,
        TOTAL_PRICE_LOWEST,
        TOTAL_PRICE_HIGHEST,
        DEPART_EARLIEST,
        DEPART_LATEST,
        AIRLINES,
        UNKNOWN
    }

    companion object {
        val instance = PriceComparator()
    }

    private var mOrder = DEPART_EARLIEST

    override fun compare(price1: PricePerDay?, price2: PricePerDay?): Int {
        if (price1 == null && price2 == null) {
            return 0
        }

        if (price1 == null) {
            return -1
        }

        if (price2 == null) {
            return 1
        }

        return when (mOrder) {
            PRICE_ONLY_LOWEST -> price1.price.compareTo(price2.price)
            PRICE_ONLY_HIGHEST -> price2.price.compareTo(price1.price)
            TOTAL_PRICE_LOWEST -> price1.priceTotal.compareTo(price2.priceTotal)
            TOTAL_PRICE_HIGHEST -> price2.priceTotal.compareTo(price1.priceTotal)
            DEPART_EARLIEST -> compareTwoNullableDate(price1.getDepartureTime(), price2.getDepartureTime())
            DEPART_LATEST -> compareTwoNullableDate(price2.getDepartureTime(), price1.getDepartureTime())
            AIRLINES -> price1.carrier.compareTo(price2.carrier)
            else -> 0
        }
    }

    private fun compareTwoNullableDate(date1: Date?, date2: Date?): Int {
        return if (date1 == null && date2 == null) {
            0
        } else if (date1 == null) {
            -1
        } else if (date2 == null) {
            1
        } else {
            date1.compareTo(date2)
        }
    }

    fun setSortOrder(order: SortOrder) {
        mOrder = order
    }
}
